File: go/parity_test.go
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

// Cross-runtime conformance, driven by the shared `test/spec/*.tsv`
// fixtures at the repo root (see ../test/AGENTS.md).
//
// `ts/test/parity.test.ts` discovers and runs the SAME files, so the two
// implementations cannot drift without one of them going red.

package tabnaschess

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strings"
	"testing"
)

const specDir = "../test/spec"

type specRow struct {
	line     int
	input    string
	expected string
	opts     string
}

// Decode the escape set used in non-JSON columns. Kept byte-identical to
// the TypeScript loader so both runtimes feed the parser the exact same
// source text.
func unescapeCol(s string) string {
	if !strings.Contains(s, `\`) {
		return s
	}
	var out strings.Builder
	for i := 0; i < len(s); i++ {
		c := s[i]
		if '\\' == c && i+1 < len(s) {
			switch s[i+1] {
			case 'n':
				out.WriteByte('\n')
				i++
				continue
			case 'r':
				out.WriteByte('\r')
				i++
				continue
			case 't':
				out.WriteByte('\t')
				i++
				continue
			case '\\':
				out.WriteByte('\\')
				i++
				continue
			}
		}
		out.WriteByte(c)
	}
	return out.String()
}

func loadSpec(t *testing.T, file string) []specRow {
	t.Helper()
	body, err := os.ReadFile(filepath.Join(specDir, file))
	if nil != err {
		t.Fatalf("cannot read %s: %v", file, err)
	}
	lines := strings.Split(strings.ReplaceAll(string(body), "\r\n", "\n"), "\n")

	rows := []specRow{}
	// Line 1 is the header naming the columns.
	for i := 1; i < len(lines); i++ {
		raw := lines[i]
		// A comment line starts with '#' and has no tab; a data row always
		// has at least one, so a '#'-leading source still works.
		if "" == raw || (strings.HasPrefix(raw, "#") && !strings.Contains(raw, "\t")) {
			continue
		}
		col := strings.Split(raw, "\t")
		if 2 > len(col) {
			t.Fatalf("%s:%d: expected at least 2 tab-separated columns", file, i+1)
		}
		row := specRow{line: i + 1, input: unescapeCol(col[0]), expected: col[1]}
		if 3 <= len(col) {
			row.opts = col[2]
		}
		rows = append(rows, row)
	}
	return rows
}

// A truncated single-line rendering of the input, so a failure names its case.
func label(s string) string {
	one := strings.ReplaceAll(s, "\n", " ; ")
	if 60 < len(one) {
		return one[:57] + "..."
	}
	return one
}

// optionsFromJSON reads the fixture's `opts` column. The column is the
// TypeScript plugin's option object, so the names are the TS ones.
func optionsFromJSON(t *testing.T, raw string) Options {
	t.Helper()
	if "" == strings.TrimSpace(raw) {
		return Options{}
	}
	var m map[string]any
	if err := json.Unmarshal([]byte(raw), &m); nil != err {
		t.Fatalf("bad opts %q: %v", raw, err)
	}
	return optionsFromMap(m)
}

// parseWith runs the configured start rule and returns the result as the
// generic JSON shape the fixture states, so the comparison is against what
// a consumer actually receives.
func parseWith(o Options, src string) (any, error) {
	j := Make(o)
	out, err := j.Parse(src)
	if nil != err {
		return nil, err
	}
	encoded, err := json.Marshal(out)
	if nil != err {
		return nil, err
	}
	var generic any
	if err := json.Unmarshal(encoded, &generic); nil != err {
		return nil, err
	}
	return generic, nil
}

func TestSpec(t *testing.T) {
	entries, err := os.ReadDir(specDir)
	if nil != err {
		t.Fatalf("cannot read %s: %v", specDir, err)
	}

	files := []string{}
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".tsv") {
			files = append(files, e.Name())
		}
	}
	sort.Strings(files)
	if 0 == len(files) {
		t.Fatalf("no .tsv fixtures in %s", specDir)
	}

	for _, file := range files {
		t.Run("spec: "+file, func(t *testing.T) {
			rows := loadSpec(t, file)
			if 0 == len(rows) {
				t.Fatalf("%s: no cases", file)
			}
			for _, row := range rows {
				t.Run(label(row.input), func(t *testing.T) {
					o := optionsFromJSON(t, row.opts)
					got, err := parseWith(o, row.input)

					if strings.HasPrefix(row.expected, "ERROR") {
						want := strings.TrimPrefix(
							strings.TrimPrefix(row.expected, "ERROR"), ":")
						if nil == err {
							t.Fatalf("%s:%d: expected %s, got %v",
								file, row.line, row.expected, got)
						}
						if "" != want && !strings.Contains(err.Error(), want) {
							t.Fatalf("%s:%d: expected error containing %q, got %v",
								file, row.line, want, err)
						}
						return
					}

					if nil != err {
						t.Fatalf("%s:%d: %v", file, row.line, err)
					}

					var want any
					if err := json.Unmarshal([]byte(row.expected), &want); nil != err {
						t.Fatalf("%s:%d: bad expected JSON: %v", file, row.line, err)
					}
					if !reflect.DeepEqual(got, want) {
						gotJSON, _ := json.Marshal(got)
						t.Fatalf("%s:%d:\n got: %s\nwant: %s",
							file, row.line, gotJSON, row.expected)
					}
				})
			}
		})
	}
}

File: rs/tests/parity_test.rs
/* Copyright (c) 2026 Richard Rodger and other contributors, MIT License */

//! Cross-runtime conformance, driven by the shared `test/spec/*.tsv`
//! fixtures at the repo root (see ../../test/AGENTS.md).
//!
//! `ts/test/parity.test.ts` and `go/parity_test.go` discover and run the
//! SAME files, so no two of the three implementations can drift without
//! one of them going red.

use std::fs;
use std::path::{Path, PathBuf};

use tabnas::Tabnas;
use tabnas_chess::{chess, to_json, ChessOptions};

const SPEC_DIR: &str = "../test/spec";

struct SpecRow {
    line: usize,
    input: String,
    expected: String,
    opts: String,
}

/// Decode the escape set used in non-JSON columns. Kept byte-identical to
/// the TypeScript and Go loaders so all three runtimes feed the parser the
/// exact same source text.
fn unescape_col(column: &str) -> String {
    if !column.contains('\\') {
        return column.to_string();
    }
    let mut out = String::with_capacity(column.len());
    let bytes = column.as_bytes();
    let mut at = 0;
    while at < bytes.len() {
        if b'\\' == bytes[at] && at + 1 < bytes.len() {
            let decoded = match bytes[at + 1] {
                b'n' => Some('\n'),
                b'r' => Some('\r'),
                b't' => Some('\t'),
                b'\\' => Some('\\'),
                _ => None,
            };
            if let Some(decoded) = decoded {
                out.push(decoded);
                at += 2;
                continue;
            }
        }
        // `column` is UTF-8 and every byte matched above is ASCII, so the
        // remainder is copied a character at a time rather than a byte.
        let character = column[at..].chars().next().expect("index is a boundary");
        out.push(character);
        at += character.len_utf8();
    }
    out
}

fn load_spec(file: &Path) -> Vec<SpecRow> {
    let body = fs::read_to_string(file)
        .unwrap_or_else(|error| panic!("cannot read {}: {error}", file.display()));
    let body = body.replace("\r\n", "\n");

    let mut rows = Vec::new();
    // Line 1 is the header naming the columns.
    for (index, raw) in body.split('\n').enumerate().skip(1) {
        // A comment line starts with '#' and has no tab; a data row always
        // has at least one, so a '#'-leading source still works.
        if raw.is_empty() || (raw.starts_with('#') && !raw.contains('\t')) {
            continue;
        }
        let column: Vec<&str> = raw.split('\t').collect();
        assert!(
            2 <= column.len(),
            "{}:{}: expected at least 2 tab-separated columns",
            file.display(),
            index + 1
        );
        rows.push(SpecRow {
            line: index + 1,
            input: unescape_col(column[0]),
            expected: column[1].to_string(),
            opts: column.get(2).unwrap_or(&"").to_string(),
        });
    }
    rows
}

/// A truncated single-line rendering of the input, so a failure names its
/// case.
fn label(input: &str) -> String {
    let one = input.replace('\n', " ; ");
    if 60 < one.chars().count() {
        let head: String = one.chars().take(57).collect();
        return format!("{head}...");
    }
    one
}

/// Read the fixture's `opts` column. The column is the TypeScript
/// plugin's option object, so the names are the TS ones.
fn options_from_json(raw: &str) -> ChessOptions {
    if raw.trim().is_empty() {
        return ChessOptions::default();
    }
    serde_json::from_str(raw).unwrap_or_else(|error| panic!("bad opts {raw:?}: {error}"))
}

/// Run the configured start rule and return the result as the generic
/// JSON shape the fixture states, so the comparison is against what a
/// consumer actually receives.
fn parse_with(options: &ChessOptions, src: &str) -> Result<serde_json::Value, String> {
    let mut tn = Tabnas::new();
    // Fixed, rather than left to `make`: a fixture compares text, and an
    // escape code from an incidental terminal would change it.
    tn.options.color.active = false;
    chess(&mut tn, options).expect("the chess grammar installs");
    tn.parse(src)
        .map(|value| to_json(&value))
        .map_err(|error| error.to_string())
}

fn spec_files() -> Vec<PathBuf> {
    let mut files: Vec<PathBuf> = fs::read_dir(SPEC_DIR)
        .unwrap_or_else(|error| panic!("cannot read {SPEC_DIR}: {error}"))
        .map(|entry| entry.expect("a readable directory entry").path())
        .filter(|path| Some("tsv") == path.extension().and_then(|kind| kind.to_str()))
        .collect();
    files.sort();
    assert!(!files.is_empty(), "no .tsv fixtures in {SPEC_DIR}");
    files
}

#[test]
fn spec() {
    let mut failures: Vec<String> = Vec::new();
    let mut cases = 0;

    for file in spec_files() {
        let name = file
            .file_name()
            .expect("a file")
            .to_string_lossy()
            .to_string();
        let rows = load_spec(&file);
        assert!(!rows.is_empty(), "{name}: no cases");

        for row in rows {
            cases += 1;
            let options = options_from_json(&row.opts);
            let at = format!("{name}:{} [{}]", row.line, label(&row.input));
            let got = parse_with(&options, &row.input);

            if let Some(want) = row.expected.strip_prefix("ERROR") {
                let want = want.strip_prefix(':').unwrap_or(want);
                match got {
                    Ok(value) => {
                        failures.push(format!("{at}: expected {}, got {value}", row.expected))
                    }
                    Err(error) if !want.is_empty() && !error.contains(want) => failures.push(
                        format!("{at}: expected error containing {want:?}, got {error}"),
                    ),
                    Err(_) => {}
                }
                continue;
            }

            let want: serde_json::Value = serde_json::from_str(&row.expected)
                .unwrap_or_else(|error| panic!("{at}: bad expected JSON: {error}"));
            match got {
                Err(error) => failures.push(format!("{at}: {error}")),
                Ok(value) if value != want => {
                    failures.push(format!("{at}:\n got: {value}\nwant: {want}"))
                }
                Ok(_) => {}
            }
        }
    }

    println!("ran {cases} spec cases across the shared fixtures");
    assert!(
        failures.is_empty(),
        "{} of {cases} spec cases failed:\n\n{}",
        failures.len(),
        failures.join("\n\n")
    );
}
